File: src/main/java/org/kite9/diagram/common/algorithms/fg/StateStorage.java
package org.kite9.diagram.common.algorithms.fg;

import java.util.HashMap;
import java.util.Map;

/**
 * Allows you to store and restore the state of a flow graph.
 * 
 * @author robmoffat
 *
 */
public class StateStorage {

	public static Map<Object, Integer> storeState(FlowGraph fg) {
		Map<Object, Integer> state = new HashMap<Object, Integer>(100);
		for (Node n : fg.getAllNodes()) {
			state.put(n, n.getFlow());
		}

		for (Arc a : fg.getAllArcs()) {
			state.put(a, a.getFlow());
		}
		
		return state;
	}
	
	public static void restoreState(FlowGraph fg, Map<Object, Integer> state) {
		for (Node n : fg.getAllNodes()) {
			int flow = state.get(n);
			n.setFlow(flow);
			n.setSupply(-flow);
		}

		for (Arc a : fg.getAllArcs()) {
			Integer val = state.get(a);
			if (val!=null)
				a.setFlow(val);
		}
	}
}

File: src/main/kotlin/org/kite9/diagram/common/algorithms/fg/StateStorage.kt
package org.kite9.diagram.common.algorithms.fg


/**
 * Allows you to store and restore the state of a flow graph.
 *
 * @author robmoffat
 */
object StateStorage {

    @JvmStatic
	fun storeState(fg: FlowGraph): Map<Any, Int> {
        val state: MutableMap<Any, Int> = HashMap(100)
        for (n in fg.allNodes) {
            state[n] = n.flow
        }
        for (a in fg.allArcs) {
            state[a] = a.flow
        }
        return state
    }

    @JvmStatic
	fun restoreState(fg: FlowGraph, state: Map<Any, Int>) {
        for (n in fg.allNodes) {
            val flow = state[n]!!
            n.flow = flow
            n.supply = -flow
        }
        for (a in fg.allArcs) {
            val `val` = state[a]
            if (`val` != null) a.flow = `val`
        }
    }
}